addons/survey/controllers/survey_session_manage.py

```diff
@@ -19,21 +19,29 @@ def _fetch_from_token(self, survey_token):
         return request.env['survey.survey'].search([('access_token', '=', survey_token)])
 
     def _fetch_from_session_code(self, session_code):
-        """ Matches a survey against a passed session_code.
+        """ Matches a survey against a passed session_code, and checks if it is valid.
+        If it is valid, returns the start url. Else, the error type.
         We don't limit the session_state to be reachable (ready / in_progress) here because
-        in some cases, we want closed session as well (where session_state = False).
+        in some cases, we want not yet launched and closed sessions as well (where session_state = False).
         Instead, when necessary, the reachability is forced in routes calling this method to
-        avoid people using those routes to access other (private) surveys.
-        We limit to sessions opened within the last 7 days to avoid potential abuses. """
+        avoid people using those routes to access other (private) surveys."""
         if session_code:
-            matching_survey = request.env['survey.survey'].sudo().search([
-                ('session_start_time', '>', fields.Datetime.now() - relativedelta(days=7)),
-                ('session_code', '=', session_code),
-            ], limit=1)
-            if matching_survey:
-                return matching_survey
-
-        return False
+            survey = request.env['survey.survey'].sudo().search([('session_code', '=', session_code)], limit=1)
+            if survey:
+                if survey.session_state in ['ready', 'in_progress']:
+                    return {'survey_url': '/survey/start/%s' % survey.access_token}
+                elif survey.session_start_time:
+                    # Limit to surveys closed max 7 days ago
+                    if survey.session_start_time < fields.Datetime.now() - relativedelta(days=7):
+                        return {'error': 'survey_wrong'}
+                    else:
+                        return {'error': 'survey_session_closed'}
+                else:
```

Comment:
Since this added information is only useful for the survey users, I would add a condition here.

`elif request.env.user.has_group("survey.group_survey_user")`.

That avoids people trying to fetch information from private surveys.